Modules/modFileEncoding.ACM

```diff
@@ -0,0 +1,437 @@
+  Dim f As Long
+  Dim byts() As Byte
+  
+  If FileExists(strFilename) Then
+    ' load the file into the bytsFile array
+    f = FreeFile()
+    Open strFilename For Binary Access Read As #f
+    If LOF(f) >= 2 Then
+      ReDim byts(1)
+      Get #f, 1, byts()
+      IsUCS2 = ((byts(0) = 255) And (byts(1) = 254)) ' &HFF and &HFE
+    End If
+    Close #f
+  End If
+End Function
+
+Function IsUTF8(strFilename As String) As Boolean
+  ' Returns true if the file specified is in UTF8 format
+  ' NB this needs the BOM to be present in order to identify as UTF8 - this is not strictly in accordance with
+  ' UTF8 specififcations, but is the same as many Windows implementations
+  Dim f As Long
+  Dim byts() As Byte
+  
+  If FileExists(strFilename) Then
+    ' load the file into the bytsFile array
+    f = FreeFile()
+    Open strFilename For Binary Access Read As #f
+    If LOF(f) >= 3 Then
+      ReDim byts(2)
+      Get #f, 1, byts()
+      IsUTF8 = ((byts(0) = 239) And (byts(1) = 187) And (byts(1) = 191)) '&HEF, &HBB, &HBF
+    End If
+    Close #f
+  End If
+End Function
+
+Function ConvFile_to_UCS2(strFilename As String, Optional strNewFilename As String) As Boolean
+  ' Converts the supplied file from any of ANSI, UTF-8 or UCS2 into UCS2 format
+  
+  ConvFile_to_UCS2 = False ' initially set to false
+
+  If LenB(strNewFilename) = 0 Then strNewFilename = strFilename
+
+  ' deal with case where no conversino is required
+  If IsUCS2(strFilename) Then
+    If strFilename = strNewFilename Then
+      ' do nothing the file is already in UCS2 format
+      ConvFile_to_UCS2 = True
+    Else
+      ' copy file to new location without conversion
+      If FileExists(strNewFilename) Then Kill strNewFilename
+      FileCopy strFilename, strNewFilename
+      ConvFile_to_UCS2 = True
+    End If
+    Exit Function
+  End If
+  
+  ' we need to do a conversion
+  If IsUTF8(strFilename) Then
+    ConvFile_to_UCS2 = ConvFileUTF8_to_UCS2(strFilename, strNewFilename)
+  ElseIf IsANSI(strFilename) Then
+    ConvFile_to_UCS2 = ConvFileANSI_to_UCS2(strFilename, strNewFilename)
+  End If
+  
+End Function
+
+Function ConvFileANSI_to_UCS2(strFilename As String, Optional strNewFilename As String) As Boolean
+  ' Converts the supplied file from ANSI (8 bit) to UCS2 format
+  
+  Dim byts() As Byte
+  Dim bytsOut() As Byte
+  Dim lngLen As Long
+  Dim lngOutLen As Long
+  Dim i As Long
+  Dim j As Long
+    
+  ConvFileANSI_to_UCS2 = False ' initially set to false
+
+  If LenB(strNewFilename) = 0 Then strNewFilename = strFilename
+
+  If ReadFromFile(strFilename, byts) Then
+    lngLen = UBound(byts) + 1
+    lngOutLen = lngLen * 2 + 2 ' 2 bytes for every ANSI byte plus 2 Byte BOM
+    ReDim bytsOut(lngOutLen - 1)
+    ' Insert BOM
+    bytsOut(0) = 255 ' &HFF
+    bytsOut(1) = 254 ' &HFE
+    ' j is the position in the output array
+    j = 2
+    For i = 0 To UBound(byts)
+      bytsOut(j) = byts(i)

```

Comment:
Does not handle codepoints in the `80` to `9F` range. These are different in the ANSI and Unicode encodings

Reply:
Lookup array added to deal with these codepoints